tests/e2e/lib/virt_helpers.go

```diff
@@ -0,0 +1,408 @@
+	if err != nil {
+		log.Printf("Failed to create subscription: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// Creates a HyperConverged Operator instance. Another dynamic client to avoid
+// bringing in the KubeVirt APIs for now.
+func (v *VirtOperator) installHco() error {
+	resourceId := schema.GroupVersionResource{
+		Group:    "hco.kubevirt.io",
+		Resource: "hyperconvergeds",
+		Version:  "v1beta1",
+	}
+
+	unstructuredHco := unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "hco.kubevirt.io/v1beta1",
+			"kind":       "HyperConverged",
+			"metadata": map[string]interface{}{
+				"name":      "kubevirt-hyperconverged",
+				"namespace": v.Namespace,
+			},
+			"spec": map[string]interface{}{},
+		},
+	}
+
+	_, err := v.Dynamic.Resource(resourceId).Namespace(v.Namespace).Create(context.Background(), &unstructuredHco, v1.CreateOptions{})
+	if err != nil {
+		log.Printf("Error creating HCO: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// Helper function to poll status with a timeout
+func waitForCheck(timeout, period time.Duration, checkFunction func() bool, errorMessage string) error {
+	timer := time.After(timeout)
+	for {
+		if check := checkFunction(); check {
+			break
+		}
+		select {
+		case <-timer:
+			return errors.New(errorMessage)
+		default:
+			time.Sleep(period)
+		}
+	}
+	return nil
+}
+
+// Creates target namespace if needed, and waits for it to exist
+func (v *VirtOperator) ensureNamespace(timeout time.Duration) error {
+	if !v.checkNamespace() {
+		if err := v.installNamespace(); err != nil {
+			return err
+		}
+		if err := waitForCheck(timeout, 1*time.Second, v.checkNamespace, "timed out waiting to create namespace "+v.Namespace); err != nil {
+			return err
+		}
+	} else {
+		log.Printf("Namespace %s already present, no action required", v.Namespace)
+	}
+
+	return nil
+}
+
+// Creates operator group if needed, and waits for it to exist
+func (v *VirtOperator) ensureOperatorGroup(timeout time.Duration) error {
+	if !v.checkOperatorGroup() {
+		if err := v.installOperatorGroup(); err != nil {
+			return err
+		}
+		if err := waitForCheck(timeout, 1*time.Second, v.checkOperatorGroup, "timed out waiting to create operator group kubevirt-hyperconverged-group"); err != nil {
+			return err
+		}
+	} else {
+		log.Printf("Operator group already present, no action required")
+	}
+
+	return nil
+}
+
+// Creates the virtualization subscription if needed, and waits for it to exist
+func (v *VirtOperator) ensureSubscription(timeout time.Duration) error {
+	if !v.checkSubscription() {
+		if err := v.installSubscription(); err != nil {
+			return err
+		}
+		if err := waitForCheck(timeout, 1*time.Second, v.checkSubscription, "timed out waiting to create subscription"); err != nil {
+			return err
+		}
+	} else {
+		log.Printf("Subscription already created, no action required")
+	}
+
+	return nil
+}
+
+// Waits for the ClusterServiceVersion to go to ready, triggered by subscription
```

Comment:
Is that required? Why not to wait for csv when the csv is being created?

Reply:
I don't know what is really required or not here. It's just the way I found to get the operator installed, directly translated from the command line version.

Reply:
I am fine waiting here, just move it maybe inside the subscription function in it's body rather then separate function outside of the subscription. This way user will always wait for it to be ready. Otherwise consumer of function which creates subscription will need to remember to also wait for csv.

Reply:
I think that would be a good plan for a generic operator installer module, but here the only public interface is IsVirtInstalled and EnsureVirtInstallation. A consumer of these should not have to think about subscriptions, CSVs, or HCOs at all.

These could certainly be combined, but I am slightly more in favor of keeping it split into smaller chunks. If for whatever reason the Subscription creation failed, having them combined would mean you would have to wait for the longer timeout (5 minutes vs. 10 seconds) to fail to test. That's a small and unlikely failure mode, but enough of a reason to demand further convincing in favor of combining them.

Reply:
If we have long running concurrent operations then it makes sense to first ask for all of them and then check if we got them independently, so we save some blocking calls. But I am fine having a split here as well.